tasks/ci.rake

```diff
@@ -0,0 +1,283 @@
+
+OPTIONAL
+--------
+
+HOSTS:
+    The hosts to run against. Must be specified as a path to a file or a
+    beaker-hostgenerator string. Defaults to #{DEFAULT_TEST_TARGETS}.
+
+    HOSTS=mynodes.yaml
+    HOSTS=redhat7-64ma
+
+TESTS:
+    A comma-delimited list of files/directories containing tests to run.
+    Defaults to running all tests, except for `ci:test:quick` which defaults
+    to running a predefined set of tests.
+
+    TESTS=tests/cycle_detection.rb
+    TESTS=tests/pluginsync,tests/language
+
+OPTIONS:
+    Additional options to pass to beaker. Defaults to ''.
+
+    OPTIONS='--dry-run --no-color'
+
+BEAKER_HOSTS:
+    Same as HOSTS.
+
+TEST_TARGET:
+    A beaker-hostgenerator string of agent-only hosts to run against.
+    The MASTER_TEST_TARGET host will be added this list. This option is
+    only intended to be used in CI.
+
+MASTER_TEST_TARGET:
+    Override the default master test target. Should only be used with
+    TEST_TARGET, and is only intended to be used in CI. Defaults to
+    #{DEFAULT_MASTER_TEST_TARGET}.
+EOS
+
+namespace :ci do
+  desc "Print usage information"
+  task :help do
+    puts USAGE
+    exit 1
+  end
+
+  task :check_env do
+    if ENV['SHA'].nil?
+      puts "Error: A SHA must be specified"
+      puts "\n"
+      puts USAGE
+      exit 1
+    end
+
+    if ENV['TESTS'].nil?
+      ENV['TESTS'] ||= ENV['TEST']
+      ENV['TESTS'] ||= 'tests'
+    end
+  end
+
+  task :gen_hosts do
+    hosts =
+      if ENV['HOSTS']
+        ENV['HOSTS']
+      elsif ENV['BEAKER_HOSTS']
+        ENV['BEAKER_HOSTS']
+      elsif env_config = ENV['CONFIG']
+        puts 'Warning: environment variable CONFIG deprecated. Please use HOSTS to match beaker options.'
+        env_config
+      else
+        # By default we assume TEST_TARGET is an agent-only string
+        if agent_target = ENV['TEST_TARGET']
+          master_target = ENV['MASTER_TEST_TARGET'] || DEFAULT_MASTER_TEST_TARGET
+          "#{master_target}-#{agent_target}"
+        else
+          DEFAULT_TEST_TARGETS
+        end
+      end
+
+    if File.exists?(hosts)
+      ENV['HOSTS'] = hosts
+    else
+      hosts_file = "tmp/#{hosts}-#{SecureRandom.uuid}.yaml"
+      cli = BeakerHostGenerator::CLI.new([hosts, '--disable-default-role', '--osinfo-version', '1'])
+      FileUtils.mkdir_p('tmp') # -p ignores when dir already exists
+      File.open(hosts_file, 'w') do |fh|
+        fh.print(cli.execute)
+      end
+      ENV['HOSTS'] = hosts_file
+    end
+  end
+
+  namespace :test do
+    desc <<-EOS
+Run a limited but representative subset of acceptance tests against puppet-agent
+(AIO) packages. This task is intended to reduce testing time on a per-commit
+basis.
+
+  $ SHA=<full sha> bundle exec rake ci:test:quick
+
+SHA should be the full SHA for the puppet-agent package.
+EOS
+    task :quick => ['ci:check_env', 'ci:gen_hosts'] do
+      ENV['TESTS'] = get_test_sample.join(",")
+      Rake::Task["ci:test:aio"].invoke
+    end
```

Comment:
The `ci:test:quick` task is specific to the puppet component, e.g. the `get_test_sample` method returns tests that only live in puppet. How about keeping this task (and related helper methods) in puppet's Rakefile?